packages/compiler/auto.ts

```diff
@@ -0,0 +1,306 @@
+import * as t from '@babel/types';
+import { catchError, logIgnore, logImprovement } from './utils/log';
+import { addImport } from './utils/mod';
+import type { Info } from './babel';
+import type { NodePath } from '@babel/core';
+import { isAttributeUnsupported, isComponent } from './utils/jsx';
+import { transformBlock } from './block';
+
+// TODO: for rsc, need to find children files without 'use client' but are client components
+// https://nextjs.org/docs/app/building-your-application/rendering/client-components#using-client-components-in-nextjs
+//
+// TODO: add support for <For> for .maps
+export const parseAuto = (
+  path: NodePath<t.FunctionDeclaration | t.VariableDeclarator>,
+  info: Info
+) => {
+  if (!info.options.auto) return;
+
+  const rawComponent = path.node;
+
+  const isInProgramScope =
+    t.isIdentifier(rawComponent.id) &&
+    info.programPath.scope.hasOwnBinding(rawComponent.id.name);
+
+  if (!isInProgramScope) return;
+
+  if (
+    t.isCallExpression(path.parent) &&
+    t.isIdentifier(path.parent.callee, { name: 'block' }) // probably need a import level check later
+  )
+    return;
+
+  let componentPath: NodePath<
+    t.FunctionDeclaration | t.FunctionExpression | t.ArrowFunctionExpression
+  >;
+
+  // Next.js React Server Components support
+  if (typeof info.options.auto === 'object' && info.options.auto.rsc) {
+    const directives = info.programPath.node.directives;
+    if (!directives.length) return; // we assume it's server component only
+    const hasUseClientDirective = directives.some((directive) => {
+      return directive.value.value === 'use client';
+    });
+    if (!hasUseClientDirective) return;
+  }
+
+  if (t.isVariableDeclarator(rawComponent)) {
+    if (
+      !t.isArrowFunctionExpression(rawComponent.init) &&
+      !t.isFunctionExpression(rawComponent.init)
```

Comment:
Tip for the future: cover cases where expression is wrapped in a TS expression (e.g. `x as MyType` which is the same as `x`)